openfl-tutorials/experimental/106_Scikit_Learn_Linear_Regression_Workflow.ipynb

```diff
@@ -0,0 +1,572 @@
+    "    \n",
+    "# # Setup participants\n",
+    "aggregator = Aggregator()\n",
+    "# aggregator.private_attributes = {}\n",
+    "collaborators = []\n",
+    "collaborator_names = ['Portland', 'Seattle', 'Chandler','Bangalore']\n",
+    "for idx, collaborator_name in enumerate(collaborator_names):\n",
+    "    collaborators.append(\n",
+    "        Collaborator(\n",
+    "            name=collaborator_name, num_cpus=0, num_gpus=0.3,\n",
+    "            private_attributes_callable=callable_to_initialize_collaborator_private_attributes,\n",
+    "            n_collaborators=len(collaborator_names), index=idx, train_dataset=(X_train, Y_train),\n",
+    "            test_dataset=(X_test, Y_test), batch_size=32\n",
+    "        )\n",
+    "    )\n",
+    "local_runtime = LocalRuntime(aggregator=aggregator, collaborators=collaborators, backend='single_process')\n",
+    "print(f'Local runtime collaborators = {local_runtime.collaborators}')\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a5cf193b-62a2-403a-96b1-5fa716d9087f",
+   "metadata": {},
+   "source": [
+    "## Federated Learning Helper Functions\n",
+    "Define helper functions for training and validating the federated models."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "42187367-6a96-4d78-a576-f7154e8f987f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Define a class to split the dataset into shards\n",
+    "class ShardSplitter:\n",
+    "    def __init__(self, num_shards):\n",
+    "        self.num_shards = num_shards\n",
+    "\n",
+    "    def split(self, X, y):\n",
+    "        \"\"\"Split the given 2D numpy arrays X and y into equal shards and return list of indexes for each shard.\"\"\"\n",
+    "        num_samples = X.shape[0]\n",
+    "        shard_size = num_samples // self.num_shards\n",
+    "        indexes = np.arange(num_samples)\n",
+    "        #np.random.shuffle(indexes)\n",
+    "        \n",
+    "        shards = []\n",
+    "        for i in range(self.num_shards):\n",
+    "            start_idx = i * shard_size\n",
+    "            if i == self.num_shards - 1:\n",
+    "                # Include any remaining samples in the last shard\n",
+    "                end_idx = num_samples\n",
+    "            else:\n",
+    "                end_idx = start_idx + shard_size\n",
+    "            shards.append(indexes[start_idx:end_idx])\n",
+    "        \n",
+    "        return shards"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "a915fa13-fc10-4b02-808c-ad2c05cee297",
+   "metadata": {},
+   "source": [
+    "## Define Federated Averaging Method\n",
+    "The FedAvg method is used to average the models from all the collaborators after training."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "65e2b0a2-5a79-4f92-b255-4c8d3e28b635",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def inference(model, test_loader, batch_size):\n",
```

Comment:
remove this method, we are not using this.

Reply:
changes done in 26834c40e44d73315398addd340c2138cf4159b7